File: Sources/Image/UIImage/UIImage+ImageEffects.m
//
//  UIImage+ImageEffects.m
//  Any
//
//  Created by ZJaDe on 2018/6/20.
//  Copyright © 2018年 ZJaDe. All rights reserved.
//

#import "UIImage+ImageEffects.h"
#import <float.h>
@import Accelerate;

@implementation UIImage (ImageEffects)

- (UIImage* _Nullable)applyBlurWithRadius:(CGFloat)blurRadius
                               tintColor:(UIColor* _Nullable)tintColor
                   saturationDeltaFactor:(CGFloat)saturationDeltaFactor
                               maskImage:(UIImage* _Nullable)maskImage {
    
    // Check pre-conditions.
    if (self.size.width < 1 || self.size.height < 1) {
        NSLog (@"*** error:invalid size:(%.2f x %.2f). Both dimensions must be >= 1:%@", self.size.width, self.size.height, self);
        return nil;
    }
    
    if (!self.CGImage) {
        NSLog (@"*** error:image must be backed by a CGImage:%@", self);
        return nil;
    }
    
    if (maskImage && !maskImage.CGImage) {
        NSLog (@"*** error:maskImage must be backed by a CGImage:%@", maskImage);
        return nil;
    }
    
    CGRect   imageRect   = { CGPointZero, self.size };
    UIImage *effectImage = self;
    
    BOOL hasBlur             = blurRadius > __FLT_EPSILON__;
    BOOL hasSaturationChange = fabs(saturationDeltaFactor - 1.) > __FLT_EPSILON__;
    if (hasBlur || hasSaturationChange) {
        
        UIGraphicsBeginImageContextWithOptions(self.size, NO, [[UIScreen mainScreen] scale]);
        CGContextRef effectInContext = UIGraphicsGetCurrentContext();
        CGContextScaleCTM(effectInContext, 1.0, -1.0);
        CGContextTranslateCTM(effectInContext, 0, -self.size.height);
        CGContextDrawImage(effectInContext, imageRect, self.CGImage);
        
        vImage_Buffer effectInBuffer;
        effectInBuffer.data     = CGBitmapContextGetData(effectInContext);
        effectInBuffer.width    = CGBitmapContextGetWidth(effectInContext);
        effectInBuffer.height   = CGBitmapContextGetHeight(effectInContext);
        effectInBuffer.rowBytes = CGBitmapContextGetBytesPerRow(effectInContext);
        
        UIGraphicsBeginImageContextWithOptions(self.size, NO, [[UIScreen mainScreen] scale]);
        CGContextRef effectOutContext = UIGraphicsGetCurrentContext();
        vImage_Buffer effectOutBuffer;
        effectOutBuffer.data     = CGBitmapContextGetData(effectOutContext);
        effectOutBuffer.width    = CGBitmapContextGetWidth(effectOutContext);
        effectOutBuffer.height   = CGBitmapContextGetHeight(effectOutContext);
        effectOutBuffer.rowBytes = CGBitmapContextGetBytesPerRow(effectOutContext);
        
        if (hasBlur) {
            
            // A description of how to compute the box kernel width from the Gaussian
            // radius (aka standard deviation) appears in the SVG spec:
            // http://www.w3.org/TR/SVG/filters.html#feGaussianBlurElement
            //
            // For larger values of 's' (s >= 2.0), an approximation can be used:Three
            // successive box-blurs build a piece-wise quadratic convolution kernel, which
            // approximates the Gaussian kernel to within roughly 3%.
            //
            // let d = floor(s * 3*sqrt(2*pi)/4 + 0.5)
            //
            // ... if d is odd, use three box-blurs of size 'd', centered on the output pixel.
            //
            CGFloat inputRadius = blurRadius * [[UIScreen mainScreen] scale];
            NSUInteger radius = floor(inputRadius * 3. * sqrt(2 * M_PI) / 4 + 0.5);
            if (radius % 2 != 1) {
                
                radius += 1; // force radius to be odd so that the three box-blur methodology works.
            }
            
            vImageBoxConvolve_ARGB8888(&effectInBuffer, &effectOutBuffer, NULL, 0, 0, (uint32_t)radius, (uint32_t)radius, 0, kvImageEdgeExtend);
            vImageBoxConvolve_ARGB8888(&effectOutBuffer, &effectInBuffer, NULL, 0, 0, (uint32_t)radius, (uint32_t)radius, 0, kvImageEdgeExtend);
            vImageBoxConvolve_ARGB8888(&effectInBuffer, &effectOutBuffer, NULL, 0, 0, (uint32_t)radius, (uint32_t)radius, 0, kvImageEdgeExtend);
        }
        
        BOOL effectImageBuffersAreSwapped = NO;
        if (hasSaturationChange) {
            
            CGFloat s = saturationDeltaFactor;
            CGFloat floatingPointSaturationMatrix[] = {
                0.0722 + 0.9278 * s, 0.0722 - 0.0722 * s, 0.0722 - 0.0722 * s, 0,
                0.7152 - 0.7152 * s, 0.7152 + 0.2848 * s, 0.7152 - 0.7152 * s, 0,
                0.2126 - 0.2126 * s, 0.2126 - 0.2126 * s, 0.2126 + 0.7873 * s, 0,
                0,                   0,                   0, 1,
            };
            const int32_t divisor = 256;
            NSUInteger matrixSize = sizeof(floatingPointSaturationMatrix)/sizeof(floatingPointSaturationMatrix[0]);
            int16_t saturationMatrix[matrixSize];
            
            for (NSUInteger i = 0; i < matrixSize; ++i) {
                
                saturationMatrix[i] = (int16_t)roundf(floatingPointSaturationMatrix[i] * divisor);
            }
            
            if (hasBlur) {
                
                vImageMatrixMultiply_ARGB8888(&effectOutBuffer, &effectInBuffer, saturationMatrix, divisor, NULL, NULL, kvImageNoFlags);
                effectImageBuffersAreSwapped = YES;
                
            } else {
                
                vImageMatrixMultiply_ARGB8888(&effectInBuffer, &effectOutBuffer, saturationMatrix, divisor, NULL, NULL, kvImageNoFlags);
            }
        }
        
        if (!effectImageBuffersAreSwapped) {
            
            effectImage = UIGraphicsGetImageFromCurrentImageContext();
        }
        
        UIGraphicsEndImageContext();
        
        if (effectImageBuffersAreSwapped) {
            
            effectImage = UIGraphicsGetImageFromCurrentImageContext();
        }
        
        UIGraphicsEndImageContext();
    }
    
    // Set up output context.
    UIGraphicsBeginImageContextWithOptions(self.size, NO, [[UIScreen mainScreen] scale]);
    CGContextRef outputContext = UIGraphicsGetCurrentContext();
    CGContextScaleCTM(outputContext, 1.0, -1.0);
    CGContextTranslateCTM(outputContext, 0, -self.size.height);
    
    // Draw base image.
    CGContextDrawImage(outputContext, imageRect, self.CGImage);
    
    // Draw effect image.
    if (hasBlur) {
        
        CGContextSaveGState(outputContext);
        
        if (maskImage) {
            
            CGContextClipToMask(outputContext, imageRect, maskImage.CGImage);
        }
        
        CGContextDrawImage(outputContext, imageRect, effectImage.CGImage);
        CGContextRestoreGState(outputContext);
    }
    
    // Add in color tint.
    if (tintColor) {
        
        CGContextSaveGState(outputContext);
        CGContextSetFillColorWithColor(outputContext, tintColor.CGColor);
        CGContextFillRect(outputContext, imageRect);
        CGContextRestoreGState(outputContext);
    }
    
    // Output image is ready.
    UIImage *outputImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return outputImage;
}


- (UIColor * _Nullable)averageColor {
    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
    unsigned char rgba[4];
    CGContextRef context = CGBitmapContextCreate(rgba, 1, 1, 8, 4, colorSpace, kCGImageAlphaPremultipliedLast | kCGBitmapByteOrder32Big);
    
    CGContextDrawImage(context, CGRectMake(0, 0, 1, 1), self.CGImage);
    CGColorSpaceRelease(colorSpace);
    CGContextRelease(context);
    
    if(rgba[3] > 0) {
        CGFloat alpha = ((CGFloat)rgba[3])/255.0;
        CGFloat multiplier = alpha/255.0;
        return [UIColor colorWithRed:((CGFloat)rgba[0])*multiplier
                               green:((CGFloat)rgba[1])*multiplier
                                blue:((CGFloat)rgba[2])*multiplier
                               alpha:alpha];
    }
    else {
        return [UIColor colorWithRed:((CGFloat)rgba[0])/255.0
                               green:((CGFloat)rgba[1])/255.0
                                blue:((CGFloat)rgba[2])/255.0
                               alpha:((CGFloat)rgba[3])/255.0];
    }
}

- (UIColor* _Nullable)mostColor {
#if __IPHONE_OS_VERSION_MAX_ALLOWED > __IPHONE_6_1
    int bitmapInfo = kCGBitmapByteOrderDefault | kCGImageAlphaPremultipliedLast;
#else
    int bitmapInfo = kCGImageAlphaPremultipliedLast;

#endif
    //第一步 先把图片缩小 加快计算速度. 但越小结果误差可能越大
    CGSize thumbSize = CGSizeMake(self.size.width / 2, self.size.height / 2);
    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
    CGContextRef context = CGBitmapContextCreate(NULL, thumbSize.width, thumbSize.height, 8, thumbSize.width*4, colorSpace, bitmapInfo);
    CGRect drawRect = CGRectMake(0, 0, thumbSize.width/2, thumbSize.height/2);
    CGContextDrawImage(context, drawRect, self.CGImage);
    CGColorSpaceRelease(colorSpace);
    //第二步 取每个点的像素值
    unsigned char* data = CGBitmapContextGetData(context);
    if (data == NULL) return nil;
    NSCountedSet *cls=[NSCountedSet setWithCapacity:thumbSize.width*thumbSize.height];
    for (int x=0; x<thumbSize.width; x++) {
        for (int y=0; y<thumbSize.height; y++) {
            int offset = (x * y) * 4;
            int red = data[offset];
            int green = data[offset+1];
            int blue = data[offset+2];
            int alpha =  data[offset+3];
            if (alpha>0) {//去除透明
                if (red>=240&&green>=240&&blue>=240) {//去除白色
                }else{
                    NSArray *clr=@[@(red),@(green),@(blue),@(alpha)];
                    [cls addObject:clr];
                }
            }
        }
    }
    CGContextRelease(context);
    //第三步 找到出现次数最多的那个颜色
    NSEnumerator *enumerator = [cls objectEnumerator];
    NSArray *curColor = nil;
    NSArray *MaxColor=nil;
    NSUInteger MaxCount=0;
    while ( (curColor = [enumerator nextObject]) != nil ) {
        NSUInteger tmpCount = [cls countForObject:curColor];
        if ( tmpCount < MaxCount ) continue;
        MaxCount=tmpCount;
        MaxColor=curColor;
    }
    return [UIColor colorWithRed:([MaxColor[0] intValue]/255.0f) green:([MaxColor[1] intValue]/255.0f) blue:([MaxColor[2] intValue]/255.0f) alpha:([MaxColor[3] intValue]/255.0f)];
}

@end

File: Sources/Image/UIImage/UIImage+ImageEffects.swift
//
//  UIImage+ImageEffects.swift
//  Any
//
//  Created by ZJaDe on 2018/6/20.
//  Copyright © 2018年 ZJaDe. All rights reserved.
//

import UIKit

extension UIImage {
    func applyBlur(withRadius blurRadius: CGFloat,
                   tintColor: UIColor,
                   saturationDeltaFactor: CGFloat,
                   maskImage: UIImage?,
                   atFrame frame: CGRect) -> UIImage {
        if let blurredFrame = self.cropped(at: frame).applyBlur(withRadius: blurRadius, tintColor: tintColor, saturationDeltaFactor: saturationDeltaFactor, maskImage: maskImage) {
            return self.blend(blurredFrame, cropRect: frame)
        } else {
            return self
        }
    }
}

public enum ImageEffect {
    case light
    case extraLight
    case dark
    case tintEffect(UIColor)
}
extension UIImage {
    public func applyEffect(_ effect: ImageEffect) -> UIImage? {
        let tintColor: UIColor
        switch effect {
        case .light:
            tintColor = UIColor(white: 1, alpha: 0.3)
            return self.applyBlur(withRadius: 7, tintColor: tintColor, saturationDeltaFactor: 1.8, maskImage: nil)
        case .extraLight:
            tintColor = UIColor(white: 0.97, alpha: 0.82)
            return self.applyBlur(withRadius: 5, tintColor: tintColor, saturationDeltaFactor: 1.8, maskImage: nil)
        case .dark:
            tintColor = UIColor(white: 0.11, alpha: 0.73)
            return self.applyBlur(withRadius: 5, tintColor: tintColor, saturationDeltaFactor: 1.8, maskImage: nil)
        case .tintEffect(let effectColor):
            tintColor = effectColor.alpha(0.6)
            return self.applyBlur(withRadius: 5, tintColor: tintColor, saturationDeltaFactor: 1.4, maskImage: nil)
        }
    }
}
extension UIImage {
    public func blurImage(radius: CGFloat = 5, tintColor: UIColor = UIColor.clear, maskImage: UIImage? = nil) -> UIImage? {
        return self.applyBlur(withRadius: radius, tintColor: tintColor, saturationDeltaFactor: 1.4, maskImage: maskImage)
    }
    public func blurImage(at frame: CGRect) -> UIImage {
        return self.applyBlur(withRadius: 5, tintColor: UIColor.clear, saturationDeltaFactor: 1.4, maskImage: nil, atFrame: frame)
    }
}
extension UIImage {
    /// ZJaDe: 灰度图
    func grayScale() -> UIImage? {
        let width: Int = Int(self.size.width)
        let height: Int = Int(self.size.height)

        let colorSpace = CGColorSpaceCreateDeviceGray()
        guard let context = CGContext(data: nil, width: width, height: height, bitsPerComponent: 8, bytesPerRow: 0, space: colorSpace, bitmapInfo: CGImageAlphaInfo.none.rawValue) else {
            return nil
        }

        context.draw(self.cgImage!, in: CGRect(x: 0, y: 0, width: width, height: height))
        let grayImage = context.makeImage()!
        return UIImage(cgImage: grayImage)
    }
}

extension UIImage {
    /// ZJaDe: 精度无损 略微耗时
    public func mainColor() -> UIColor? {
        guard let cgImage = self.cgImage else {
            return nil
        }
        let imgWidth = cgImage.width
        let imgHeight = cgImage.height
        guard let colorSpace = cgImage.colorSpace else {
            return nil
        }
        //位图的大小＝图片宽＊图片高＊图片中每点包含的信息量
        let bitmapByteCount = imgWidth * imgHeight * 4
        guard let context = CGContext(data: nil, width: imgWidth, height: imgHeight, bitsPerComponent: 8, bytesPerRow: imgWidth * 4, space: colorSpace, bitmapInfo: cgImage.bitmapInfo.rawValue) else {
            return nil
        }
        //将图片画到位图中
        context.draw(cgImage, in: CGRect(x: 0, y: 0, width: CGFloat(imgWidth), height: CGFloat(imgHeight)))
        guard let contextData = context.data else {
            return nil
        }
        //获取位图数据
        let data = contextData.bindMemory(to: CUnsignedChar.self, capacity: bitmapByteCount)
        let mapFunc = dataFliterMap(data, imgWidth: imgWidth, imgHeight: imgHeight)
        return mapFunc(false) ?? mapFunc(true)
    }
    private func dataFliterMap(_ data: UnsafeMutablePointer<CUnsignedChar>, imgWidth: Int, imgHeight: Int) -> (Bool) -> UIColor? {
        return { (parsePureColor) in
            let colors = NSCountedSet()
            for x in 0..<imgWidth {
                for y in 0..<imgHeight {
                    let offSet = 4 * ( x * y)
                    let r = (data + offSet).pointee
                    let g = (data + offSet + 1).pointee
                    let b = (data + offSet + 2).pointee
                    let a = (data + offSet + 3).pointee
                    guard a > 0 else {
                        continue
                    }
                    ///不是纯色时或者允许解析纯色时
                    guard !(r == g && g == b) || parsePureColor else {
                        continue
                    }
                    let values = [r, g, b, a].map({CGFloat($0)})
                    colors.add(values)
                }
            }
            let orderFunc: (Any, Any) -> Bool = {colors.count(for: $0) < colors.count(for: $1)}
            guard let maxCountColor = colors.max(by: orderFunc) as? [CGFloat] else {
                return nil
            }
            //        logDebug("count-->\(colors.count(for: maxCountColor))")
            //        logDebug(colors.map({($0, colors.count(for: $0))}).sorted(by: {$0.1 < $1.1}))
            return self.mapColor(maxCountColor)
        }
    }
    private func mapColor(_ array: [CGFloat]) -> UIColor {
        return UIColor(r: array[0], g: array[1], b: array[2], a: array[3])
    }
}
